azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -176,3 +210,32 @@ def _on_message_received(self, topic, payload):
             stage=self,
             event=pipeline_events_mqtt.IncomingMQTTMessageEvent(topic=topic, payload=payload),
         )
+
+    def _on_unexpected_connection(self):
+        """
+        Handler that gets called by the transport when it connections.  This handler only
+        gets called when it connects outside of the normal connect/reconnect logic.
+        """
+        logger.warn("an unexpected connection occured")
+        self.on_connected()
+
+    def _on_unexpected_connection_failure(self, cause):
+        """
+        Handler that gets called by the transport when a connection fails.  This handler only
+        gets called when a connection fails outside of the normal connect/reconnect logic,
+        which should not happen, so the only real route of action is to pass the failure up
+        to the unhandled error handler
+        """
+        logger.error("an unexpected connection failure occured: {}".format(cause))
+        self.pipeline_root.unhandled_error_handler(cause)
+
+    def _on_unexpected_disconnection(self, cause):
+        """
+        Handler that gets called by the transport when the transport disconnects.  This handler only
+        gets called when a disconnection fails outside of the normal connect/reconnect logic,
+        which should not happen, so the only real route of action is to pass the failure up
+        to the unhandled error handler
+        """
+        logger.error("an unexpected disconnection occured: {}".format(cause))
+        self.pipeline_root.unhandled_error_handler(cause)
```

Comment:
>unhandled_error_handler [](start = 27, length = 23)

side question: I'm looking for suggestions on a better name for "unhandled_error_handler" because this should eventually bubble up to the user.  I said "unhandled" because these are errors that wouldn't be caught by any user-level try/catch logic, usually because they happen in a background thread.  What do we call these errors?